pkg/kubectl/cmd/apiresources.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/printers"
+)
+
+var (
+	apiresources_example = templates.Examples(`
+		# Print the supported API Resources
+		kubectl api-resources`)
+)
+
+// ApiResourcesOptions is the start of the data required to perform the operation.  As new fields are added, add them here instead of
+// referencing the cmd.Flags()
+type ApiResourcesOptions struct {
+	out io.Writer
+
+	APIGroup   string
+	Namespaced bool
+	NoHeaders  bool
+}
+
+// groupResource contains the APIGroup and APIResource
+type groupResource struct {
+	APIGroup    string
+	APIResource metav1.APIResource
+}
+
+func NewCmdApiResources(f cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &ApiResourcesOptions{
+		out: out,
+	}
+
+	cmd := &cobra.Command{
+		Use:     "api-resources",
+		Short:   "Print the supported API resources on the server",
+		Long:    "Print the supported API resources on the server",
+		Example: apiresources_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(options.Validate(args))
+			cmdutil.CheckErr(options.RunApiResources(cmd, f))
+		},
+	}
+	cmd.Flags().StringVar(&options.APIGroup, "api-group", "", "The API group to use when talking to the server.")
+	cmd.Flags().BoolVar(&options.Namespaced, "namespaced", true, "Namespaced indicates if a resource is namespaced or not.")
+	cmd.Flags().BoolVar(&options.NoHeaders, "no-headers", false, "When using the default or custom-column output format, don't print headers (default print headers).")
+	return cmd
+}
+
+func (o *ApiResourcesOptions) Validate(args []string) error {
+	if len(args) > 0 {
+		return fmt.Errorf("not require arguments.")
+	}
+	return nil
+}
+
+func (o *ApiResourcesOptions) RunApiResources(cmd *cobra.Command, f cmdutil.Factory) error {
+	w := printers.GetNewTabWriter(o.out)
+	defer w.Flush()
+
+	discoveryclient, err := f.DiscoveryClient()
+	if err != nil {
+		return err
+	}
+
+	lists, err := discoveryclient.ServerPreferredResources()
+	if err != nil {
```

Comment:
this can return partial results (see discovery.ErrGroupDiscoveryFailed)... it is likely still useful to display the api-resources for the other groups (but still exit with non-zero code)